libc/string_mem.c

```diff
@@ -0,0 +1,404 @@
+}
+
+
+TEST(string_memset, big)
+{
+	ret = memset(big_buf, 10, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(big_buf, ret);
+	TEST_ASSERT_EQUAL_MEMORY(expect_mem_big, big_buf, BIG_SIZE);
+}
+
+
+TEST(string_memset, single_byte)
+{
+	int8_t val = 5;
+	ret = memset(&val, 0, 1);
+	TEST_ASSERT_EQUAL_PTR(&val, ret);
+	TEST_ASSERT_EQUAL_INT8(0, val);
+}
+
+
+TEST(string_memset, byte_val)
+{
+	int i;
+	for (i = 0; i < 256; i++) {
+		memset(buf_byte + i, i, 1);
+		buf_byte_exp[i] = i;
+	}
+	TEST_ASSERT_EQUAL_MEMORY(buf_byte_exp, buf_byte, 256);
+}
+
+
+TEST_GROUP(string_memmove);
+
+
+TEST_SETUP(string_memmove)
+{
+	mem = calloc(MEM_SIZE, 1);
+	TEST_ASSERT_NOT_NULL(mem);
+	ret = memcpy(((int8_t *)mem) + OFFSET, TEST_STR, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(((int8_t *)mem) + OFFSET, ret);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem), OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem) + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST_TEAR_DOWN(string_memmove)
+{
+	free(mem);
+}
+
+
+TEST(string_memmove, no_overlap_end)
+{
+	int8_t *ptr = (int8_t *)mem + (MEM_SIZE - TEST_STR_SIZE);
+	/* Move to the end of allocated memory */
+	ret = memmove(ptr, TEST_STR, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ((int8_t *)mem) + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem) + OFFSET + TEST_STR_SIZE, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem), OFFSET);
+}
+
+
+TEST(string_memmove, no_overlap_beetwen)
+{
+	int move_offset = OFFSET + TEST_STR_SIZE + 10;
+	int8_t *ptr = (int8_t *)mem + move_offset;
+	/* Move somewhere beetwen */
+	ret = memmove(ptr, TEST_STR, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ((int8_t *)mem) + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem), OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem) + OFFSET + TEST_STR_SIZE, move_offset - (OFFSET + TEST_STR_SIZE));
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem) + move_offset + TEST_STR_SIZE, MEM_SIZE - (move_offset + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, no_overlap_adjacent)
+{
+	int8_t *ptr = (int8_t *)mem + OFFSET + TEST_STR_SIZE;
+	/* Move so that memory blocks are next to each other*/
+	ret = memmove(ptr, TEST_STR, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ((int8_t *)mem) + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem), OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, ((int8_t *)mem) + OFFSET + (TEST_STR_SIZE * 2), MEM_SIZE - (OFFSET + (TEST_STR_SIZE * 2)));
+}
+
+
+TEST(string_memmove, whole_overlap)
+{
+	int8_t *ptr = (int8_t *)mem + OFFSET;
+	/* Move at the same address as block is */
+	ret = memmove(ptr, TEST_STR, TEST_STR_SIZE);
```

Comment:
:exclamation:  these above tests do not test overlap at all. 

You need to test overlapping of the **memory** not the **values**

`<ptr, ptr + 20>` memory and `<TEST_STR, TEST_STR + 20>` memory never overlaps.

What You need to test is for example `memmove(ptr, ptr + 10, 20)` and verify if the output memory is not garbled (in `memcpy` it might as the order of copying is not defined - that's the main difference

Reply:
Thanks for catching it, we will be more careful in future tests.